src/libsystemd/sd-netlink/sd-netlink.c

```diff
@@ -892,6 +895,9 @@ int sd_netlink_add_match(
 
         LIST_PREPEND(match_callbacks, rtnl->match_callbacks, &slot->match_callback);
 
+        /* Set this at last. Otherwise, some failures in above call the destroy callback but some do not. */
+        slot->destroy_callback = destroy_callback;
```

Comment:
This redefines the semantics as "the destroy_callback" will be called if queuing this function succeeded". I think that's clear and useful.